chain/client/src/sync.rs

```diff
@@ -464,6 +421,14 @@ impl StateSync {
         }
     }
 
+    pub fn get_num_parts(&self, state_root_node: &StateRootNode) -> u64 {
+        let state_size = state_root_node.memory_usage;
+        // We assume that 1 Mb is a good limit for state part size.
+        // On the other side, it's important to divide any state into
+        // several parts to make sure that partitioning always works.
+        state_size / (1024 * 1024) + 3
```

Comment:
`+3` is to make sure that any state is divided and sent by parts. It's mostly to make sure that everything works well on small states.

> Also, could we make it a constant which we can reuse on the other side of the code that does state dividing?

Now Sync is the only source of truth how the state should be divided. I didn't get how it can be reused.

Reply:
How adding `3` makes sure that any state is divided and sent by parts and everything works well on small states?
Would `+2` or `+4` work just as well?

Reply:
> Would +2 or +4 work just as well?

Absolutely.

Reply:
> How adding 3 makes sure that any state is divided and sent by parts and everything works well on small states?

Also would `+1` work too? Would ` if state_size >= MiB {state_size / MiB } else 1` work better where `const MiB: u64 = 1024* 1024;` ?

Reply:
I don't understand what is better in our case.
The point of having this formula is described in comment:
1. Divide parts to a reasonable size (approx. up to 1 Mb)
2. Always divide the state for several parts to make sure it works after code changes or whatever.

Reply:
Wouldn't simplest formula that achieves (1) and (2) be the best choice? The simplest formula then would be `ceil(state_size, MiB)` but since Rust does not have `ceil` for non-floats its quick implementation is `state_size / MiB + if state_size % MiB == 0 {0} else {1}`

Reply:
> Also would +1 work too? 

Nope. 1 part means the whole state and no dividing.

> The simplest formula then would be ceil(state_size, MiB)

It breaks (2).

Reply:
> Nope. 1 part means the whole state and no dividing.

So if the whole state is less than `1MiB` it is not okay to not divide it? 
I don't understand why do we want to always divide state even if it is <<1MiB.
Could you rephrase or elaborate this argument, I do not understand it:
> Always divide the state for several parts to make sure it works after code changes or whatever.

Reply:
2 parts is not the same as 1 part.

1 part means whole. It will not be divided and not be collected. 2 parts mean it should be divided and collected then back.

Reply:
Let me ask another question then. During state syncing are we always going to sync state in more than one part even if state is 1KiB?

Reply:
Yes.

Reply:
For me it's also okay if we always split in debug mode and using `ceil` in release.

Reply:
So based on this comment:
>  to make sure it works after code changes or whatever.

It seems like the reason we want this to be the case is to ensure that every single test in our code uses state sync? It seems like we should have a flag `force_state_split` in some config that is only set in tests. If this is not something you want to do now please create an issue in nearcore explaining this cleanup.

Reply:
Yep, it may be a decision. What is the profit for changing the formula?

Reply:
Since there is going to be a cleanup issue there is no profit now. Please add TODO with issue number above this formula.

Reply:
Could you please create an issue and describe the problem you faced? I still don't understand what's wrong here.

Reply:
https://github.com/nearprotocol/nearcore/issues/1708